Revamp HUD code

Berthalamew opened the pull request:
- cleanup and modify most of the logic from the HudElements that was moved out to their own modules
- add structs s_new_hud_globals and s_new_hud_globals_player_info (not sure about these names)
- fix crosshair scaling code so that it doesn't down sample the bitmaps for the scaling
- add code to allow us to scale UI based on a resolution modifier (for future use)

- beneficial since the previous code had issues with resolution scaling and it uses the proper folder structures for the in-game logic
- code for resolution scaling is planned to be used in the future for a hud redraw that sako has partially started and offered to complete

Berthalamew committed:
update player_control module

- add s_player_action_weapons struct
- add static struct function to get the player control globals

Berthalamew committed:
remove HudElements module

- split up all the logic into the relevant files
- cleanup and modify most of the logic from the module that was moved out

- add structs s_new_hud_globals and s_new_hud_globals_player_info (not sure about these names)
- fix crosshair scaling code so that it doesn't down sample the bitmaps for the scaling
- add code to allow us to scale UI based on a resolution modifier (for future use)

Berthalamew committed:
add code for properly scaling motion sensor pulse

nukeulater commented on xlive/Blam/Engine/interface/first_person_camera.cpp:
```diff
@@ -0,0 +1,84 @@
+	{
+		calculated_radians_FOV = currentVariantSettings.forcedFOV * M_PI / 180.0f;
+	}
+	WriteValue(Memory::GetAddress(0x413780, 0x3B5300), calculated_radians_FOV);
```
use `WriteValue` only when writing to a write protected memory segment/page

Berthalamew replied:
Changed the following to 
*Memory::GetAddress<float*>(0x413780, 0x3B5300) = calculated_radians_FOV;

nukeulater commented on xlive/Blam/Engine/interface/first_person_camera.cpp:
```diff
@@ -0,0 +1,84 @@
+#define _USE_MATH_DEFINES
+#include <math.h>
+
+static bool b_showHUD = true;
```
we should remove all hungarian style bool prefixes

Berthalamew replied:
that bool wasn't supposed to be there but changed the name of the actual one

nukeulater commented on xlive/Blam/Engine/interface/motion_sensor.cpp:
```diff
@@ -0,0 +1,25 @@
+void motion_sensor_fix_size()
+{
+	float og_res = *Memory::GetAddress<float*>(0x3E6A9C);
+	WriteValue<float_t>(Memory::GetAddress(0x3E6A9C), og_res * SIZE_MULTIPLIER);
```
checked the code and `hud_scale` takes care of this, why is this needed?

Berthalamew replied:
This increases the resolution of the pulse so that it it will be the correct size on the screen.
If the resolution is not modified the pulse will be smaller than the radar bitmap when the proper size is used

ex. I've used a radar bitmap that's 4x bigger than before when modifying the hud_scale, the blip is still scaled 4x lower than normal
![image](https://user-images.githubusercontent.com/40469582/235472736-d3d47a90-447c-4a3e-ac0d-5b0af91714b5.png)

scaling the value properly fixes this
![image](https://user-images.githubusercontent.com/40469582/235472919-5f364566-1b50-4c7a-8206-b8f9cca8156b.png)

Berthalamew committed:
add requested changes

- change hud_upscale_size to a define
- change hud bool name
- remove WriteValue function use in hud code where applicable
- update game version to current version

Berthalamew committed:
revert back to the old crosshair size method

- the global for the crosshair size also edits the size of text on the hud. we'll have to decouple the crosshair size into a separate value going forward but we'll just edit the tags for now.

Berthalamew committed:
move highresfix out of tweaks and into the detour in hud

- remove ui_text_bounds since it's just a duplicate of rect2d

Berthalamew committed:
remove unused hud function in H2MOD

- move lens flare fix to proper module

Berthalamew committed:
fix hud hook so it executes before the WinMain function

Berthalamew committed:
separate the crosshair scale from text scale

- use a seperate float for scale inside the project and replace the reference to ours

nukeulater commented on xlive/Blam/Engine/interface/hud.cpp:
```diff
@@ -168,7 +130,11 @@ void hud_apply_pre_winmain_patches()
 	PatchCall(Memory::GetAddress(0x26406F, 0x0), update_hud_elements_display_settings_hook);
+
+	// Replace the crosshair and text scale global with our own just for the crosshair
+	WriteValue(Memory::GetAddress(0x222F9F + 4), &crosshair_size);
```
`WritePointer()` here

Berthalamew replied:
changed to WritePointer

nukeulater commented on xlive/H2MOD/Modules/Shell/Startup/WinMainH2.cpp:
```diff
@@ -94,6 +92,9 @@ int WINAPI WinMain_Halo2(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpC
 	WriteValue<BYTE>(Memory::GetAddress(0x4FAD98), 1);
 
+	// Apply patches for the hud that need to be applied before WinMain is called
+	hud_apply_pre_winmain_patches();
```
move to tweaks, there patches are applied before startup

Berthalamew replied:
moved it to tweaks

nukeulater commented on xlive/Blam/Engine/interface/hud.cpp:
```diff
@@ -168,7 +130,11 @@ void hud_apply_pre_winmain_patches()
 	p_update_hud_elements_display_settings = Memory::GetAddress<update_hud_elements_display_settings_t>(0x264A18, 0x0);
 
+	// Replace all calls to update_hud_elements_display_settings with our hook
 	PatchCall(Memory::GetAddress(0x25E1FC, 0x0), update_hud_elements_display_settings_hook);
```
extra 0x0 param is redundant here

Berthalamew replied:
removed

nukeulater commented on xlive/Blam/Engine/interface/hud.cpp:
```diff
@@ -0,0 +1,140 @@
+	return p_render_text_wrapper(a1, a2, a3, a4, a5, a6, a7, a8);
+}
+
+void set_hud_size(const float size)
```
`const` parameters should mainly be used for references or pointers to variables that shouldn't be modified

nukeulater replied:
otherwise is just extra syntax sugar

Berthalamew replied:
changed

Berthalamew committed:
add requested changes

- move hud_apply_pre_winmain_patches() function call to tweaks
- remove const from function parameters that aren't pointers
- use WritePointer instead of WriteValue for the crosshair scale replacement

Berthalamew committed:
Merge remote-tracking branch 'upstream/development-new' into hud_cleanup
